packages/flutter/lib/src/material/app_bar.dart

```diff
@@ -490,6 +491,76 @@ class _AppBarState extends State<AppBar> {
   }
 }
 
+class _FloatingAppBar extends StatefulWidget {
+  _FloatingAppBar({ Key key, this.child }) : super(key: key);
+
+  final Widget child;
+
+  @override
+  _FloatingAppBarState createState() => new _FloatingAppBarState();
+}
+
+// A wrapper for the widget created by _SliverAppBarDelegate that starts and
+/// stops the floating appbar's snap-into-view or snap-out-of-view animation.
+class _FloatingAppBarState extends State<_FloatingAppBar> {
+  ScrollableState _scrollable;
+  ScrollPosition _position;
+  ScrollActivity _activity;
+
+  @override
+  void didChangeDependencies() {
+    super.didChangeDependencies();
+    _updateScrollable();
+  }
+
+  @override
+  void dispose() {
+    _position?.removeActivityChangedListener(_scrollPositionListener);
+    super.dispose();
+  }
+
+  void _updateScrollable() {
+    _scrollable = Scrollable.of(context);
+    _updateScrollPosition();
+  }
+
+  void _updateScrollPosition() {
+    _position?.removeActivityChangedListener(_scrollPositionListener);
+    _position = _scrollable?.position;
+    _position?.addActivityChangedListener(_scrollPositionListener);
+  }
+
+  RenderSliverFloatingPersistentHeader _headerRenderer() {
+    return context.ancestorRenderObjectOfType(const TypeMatcher<RenderSliverFloatingPersistentHeader>());
+  }
+
+  void _scrollPositionListener() {
+    if (_scrollable == null)
+      return;
+
+    // When a scroll stops, then maybe snap the appbar into view.
+    // Similarly, when a scroll starts, then maybe stop the snap animation.
+    final ScrollActivity activity = _scrollable.position.activity;
+    if (_activity is! IdleScrollActivity && activity is IdleScrollActivity) {
```

Comment:
This would also let you do a level trigger rather than an edge trigger -- have the listenable be a bool value, and have the animation be on when the bool is true, rather than looking for the activity changing.